src/QuadForm/Quad/ZGenus.jl

```diff
@@ -0,0 +1,2001 @@
+    end
+    if s[4] == 0 && s[5] != 0
+      return false
+    end
+  end
+  return true
+end
+
+
+function prime(S::ZpGenus)
+  return S._prime
+end
+
+function symbol(S::ZpGenus)
+  return copy(S._symbol)
+end
+
+function genus(L::ZLat, p)
+  return genus(gram_matrix(L), p)
+end
+
+function genus(A::fmpz_mat, p, val; offset=0)
+  @assert base_ring(A)==ZZ
+  p = ZZ(p)
+  symbol = _p_adic_symbol(A, p, val)
+  for i in 1:size(symbol)[1]
+    symbol[i][1] = symbol[i][1] - offset
+  end
+  return ZpGenus(p, symbol)
+end
+
+function genus(A::MatElem, p)
+  offset = 0
+  if base_ring(A) == QQ
+    d = denominator(A)
+    val = valuation(d, p)
+    A = change_base_ring(ZZ, A*divexact(d^2, p^val))
+    offset = valuation(d, p)
+  end
+  val = valuation(det(A), p)
+  if p == 2
+    val += 3
+  end
+  return genus(A, p, val, offset=offset)
+end
+
+function Base.show(io::IO, G::ZpGenus)
+  p = G._prime
+  CS_string = ""
+  if p == 2
+    for sym in G._symbol
+      s, r, d, e, o = sym
+      d = _kronecker_symbol(d, 2)
+      if s>=0
+        CS_string *= " $(p^s)^$(d * r)"
+      else
+        CS_string *="(1/$(p^-s))^$(d * r)"
+      end
+      if e == 1
+        CS_string *= "_$o"
+      end
+    end
+  else
+    for sym in G._symbol
+      s,r ,d = sym
+      CS_string *= " $(p^s)^$(d * r)"
+    end
+  end
+  rep = "Genus symbol at $p:  $CS_string"
+  print(io, rstrip(rep))
+end
+
+function Base.:(==)(G1::ZpGenus, G2::ZpGenus)
+  # This follows p.381 Chapter 15.7 Theorem 10 in Conway Sloane's book
+  if G1._prime != G2._prime
+    raise(error("Symbols must be over the same prime to be comparable"))
+  end
+  if G1._prime != 2
+    return G1._symbol == G2._symbol
+  end
+  sym1 = symbol(G1)
+  sym2 = symbol(G2)
+  n = length(sym1)
+  @assert all(g[2]!=0 for g in sym1)
+  @assert all(g[2]!=0 for g in sym2)
+  # scales && ranks
+  s1 = [g[1:2] for g in sym1]
+  s2 = [g[1:2] for g in sym2]
+  if s1!=s2
+    return false
+  end
+  # parity
+  s1 = [g[4] for g in sym1]
+  s2 = [g[4] for g in sym2]
+  if s1 != s2
+    return false
+  end
+  push!(sym1,[sym1[end][1]+1,0,1,0,0])
+  push!(sym2,[sym1[end][1]+1,0,1,0,0])
+  prepend!(sym1,[[-1,0,1,0,0]])
+  prepend!(sym1,[[-2,0,1,0,0]])
+  prepend!(sym2,[[-1,0,1,0,0]])
+  prepend!(sym2,[[-2,0,1,0,0]])
+  n = length(sym1)
+  # oddity && sign walking conditions
+  for m in 1:n
+    # "for each integer m for which f_{2^m} has type II, we have..."
+    if sym1[m][4] == 1
+      continue # skip if type I
+    end
+    # sum_{q<2^m}(t_q-t'_q)
+    l = sum(fmpz[sym1[i][5]-sym2[i][5] for i in 1:m-1])
+    # 4 (min(a,m)+min(b,m)+...)
+    # where 2^a, 2^b are the values of q for which e_q!=e'_q
+    det_differs = [i for i in 1:n if _kronecker_symbol(sym1[i][3],2)!=_kronecker_symbol(sym2[i][3],2)]
+    r = 4*sum(fmpz[min(sym1[m][1], sym1[i][1]) for i in det_differs])
+    if 0 != mod(l-r, 8)
+      return false
+    end
+  end
+  return true
+end
+
+@doc Markdown.doc"""
+    representative(S::ZpGenus) -> MatElem
```

Comment:
For number field lattices this is called `gram_matrix` and `representative` really gives a `ZLat`.